pymeasure/instruments/instrument.py

```diff
@@ -45,6 +45,41 @@ class Instrument(object):
     :param includeSCPI: A boolean, which toggles the inclusion of standard SCPI commands
     """
 
+    class InstrumentParameter(object):
+        """ This class is used to allow the possibility to redefines 'values' parameter in property defined
+        with staticmethods 'control', 'measurement', 'setting'.
+        This class will be supported in addition to list, tuple, range and dict.
+        This class will identify the parameter by name provided as a string and the paramaters will be read from class
+        instance whenever it is needed. This will add some flexibility to modify properties in
+        subclasses without redefining the entire property.
+ 
+        :param name: A string name
+
+        .. code-block:: python
+        
+            class GenericInstrument(Instrument):
+                  __frequency_range = Instrument.InstrumentParameter("FREQUENCY_RANGE_Hz")
+                  FREQUENCY_RANGE_Hz = (1, 26.5e9) # Redefine this in subclasses to reflect actual instrument value
+
+                center_frequency = Instrument.control(
+                    ":SENS:FREQ:CENT?;", ":SENS:FREQ:CENT %e Hz;",
+                    " A floating point property that represents the frequency ... ",
+                    validator=strict_range,
+                    values=__frequency_range,
+                )
+
+            class SpecificInstrument(Instrument): # Identical to GenericInstrument, except for frequency range
+                FREQUENCY_RANGE_Hz = (1, 13e9)
+
+        """
+
+
+        def __init__(self, name):
+            self.name = name
+
+        def get_value(self, instance):
```

Comment:
Ideally a proper solution should be:
- python friendly
- not including weird syntax
- make all parameters flexible
- preserve elegance and beauty of original Instrument.control/setting/measurement mechanism
During kind of "lock-down" weekend, I come up with some other idea which is essentially subclassing the property class.
I would like to submit to evaluation and possibly add a new PR.
In order to explain it, I would prefer to add a code snippet.
```
class DynamicProperty(property):
    """ This property redefines get and set in a "dynamic" fashion """
    def __init__(self, dynlist, fget=None, fset=None, fdel=None, doc=None):
        super().__init__(fget, fset, fdel, doc)
        self.dynlist = dynlist
        self.name = ""

    def __get__(self, obj, objtype=None):
        if obj is None:
            return self
        if self.fget is None:
            raise AttributeError("unreadable attribute")

        for attr in self.dynlist:
            attr1 = "_".join([self.name,attr])
            kwargs = {}
            if hasattr(obj, attr1):
                kwargs[attr]=getattr(obj, attr1)
        return self.fget(obj, **kwargs)

    def __set__(self, obj, value):
        if self.fset is None:
            raise AttributeError("can't set attribute")
        for attr in self.dynlist:
            attr1 = "_".join([self.name,attr])
            kwargs = {}
            if hasattr(obj, attr1):
                kwargs[attr]=getattr(obj, attr1)
        self.fset(obj, value, **kwargs)

    def set_name(self, name):
        print ("Setting name: %s (old name:%s)"%(name, self.name), self)
        self.name = name

class Instrument(object):
    @staticmethod
    def property_test(values, dynamic=False, name = ""):
        param_list = ["values"] # vars().keys()
        """ Test property """
        print ("Defining property with", values)

        def fget_with_params(self, values=values):
            print ("fget: Params range is", values, self)
            return 0

        def fget(self):
            return fget_with_param(self, values)

        def fset_with_params(self, value, values=values):
            print ("fset: Params range is", values, self)

        def fset(self, value):
            return fset_with_param(self, value, values)

        if dynamic:
            return DynamicProperty(param_list, fget_with_params, fset_with_params)
        else:
            return property(fget, fset)

    def __init__(self):
        for obj in [self] + self.__class__.mro():
            for attr in obj.__dict__:
                if isinstance(obj.__dict__[attr], DynamicProperty):
                    obj.__dict__[attr].set_name(attr)

class B(Instrument):
    # Class attribute
    my_property = Instrument.property_test(values=(1,2), dynamic=True)

class C(B):
    # No inheritance, just dedicated system to set property params (<name>_<param>)
    my_property_values = (3,4)

class D(B):
    # No inheritance, just dedicated system to set property params (<name>_<param>)
    my_property_values = (5,6)

c=C()
c.my_property
d=D()
d.my_property
c.my_property
```
What do you think ?

Reply:
Regarding your comment on the get_value method, the self parameter is needed because it is an instance method (we retrieve the name of the variable controlling the parameter value) . On the other hand, the instance parameter is needed because the updated parameter value is retrieved from current instance referring the property.
Here, the idea is that there is one instance of InstrumentParameter for all the class hiererchy and instances of the class defining the property. In fact property is a class attribute and is global to all classes/subclasses and instances, so we change it by defining special name attribute (either class or instance) that alter the behaviour of fget/fset.